sycl/doc/extensions/experimental/sycl_ext_codeplay_kernel_fusion.asciidoc

```diff
@@ -0,0 +1,678 @@
+    // kernels to be fused and will not be executed before fusion is completed
+    // or cancelled.
+    q.submit([&](handler& cgh){
+      auto accIn1 = bIn1.get_access(cgh);
+      auto accIn2 = bIn2.get_access(cgh);
+      auto accTmp = bTmp.get_access(cgh);
+      cgh.parallel_for<KernelOne>(dataSize, KernelOne{accIn1, accIn2, accTmp});
+    });
+
+    // "Submit" the second kernel. The kernel will be added to the the list of
+    // kernels to be fused and will not be executed before fusion is completed
+    // or canceled.
+    q.submit([&](handler& cgh){
+      auto accTmp = bTmp.get_access(cgh);
+      auto accIn3 = bIn3.get_access(cgh);
+      auto accOut = bOut.get_access(cgh);
+      cgh.parallel_for<KernelTwo>(dataSize, KernelTwo{accTmp, accIn3, accOut});
+    });
+
+    // Complete the fusion: JIT-compile a fused kernel containing KernelOne and
+    // KernelTwo and submit the fused kernel for execution. This call may return
+    // before JIT-compilation or execution of the fused kernel is completed.
+    w.complete_fusion({ext::codeplay::property::no_barriers()});
+
+    // End of the scope - buffers go out-of-scope and are destructed. Buffer
+    // destruction causes a synchronization with all outstanding commands
+    // operating on the buffer, in this case the fused kernel.
+  }
+}
+```
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_CODEPLAY_JIT_KERNEL_FUSION` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API Extension
+
+The design tightly integrates with the `queue` class and leverages the
+asynchronous nature of SYCL kernel submissions. It introduces a new class
+`fusion_wrapper` that wraps a SYCL queue to give access to the relevant API for
+fusion. The wrapper class is introduced to achieve a separation of concerns by
+keeping the fusion control API separate from the existing queue API. The wrapper
+directly manipulates and controls the fusion state of the wrapped queue. 
+
+Next to the `fusion_wrapper`, this extension also introduces additional
+properties and a new member function for class `queue`.
+
+==== Fusion Wrapper class
+
+The `fusion_wrapper` is a thin wrapper around a SYCL queue object and provides
+access to the necessary API functions to control the fusion state of the wrapped
+queue object. The `fusion_wrapper` member functions directly modify the fusion
+state of the underlying `queue`, effectively making the queue stateful.
```

Comment:
No, this is allowed. As the state is attached to the underlying queue and not the `fusion_wrapper`, two `fusion_wrapper` objects can manage the same queue. It's the applications responsibility to synchronize if they are used in a multi-threaded context, similar to how it would be with a single `fusion_wrapper` object shared by two threads.

Reply:
OK.  Some words here explaining this would be good.

Reply:
Agreed, we have added that in the latest commit.